docs/redesign-proposal.md

```diff
@@ -0,0 +1,93 @@
+properties of how these systems work.
+
+Because the Pilot runs before the actual application has started, it is able to
+perform any preflight or configuration checks at an appropriate time. This may
+include waiting for other Pilots to start, or determining a leader (with the
+help of the `navigator-apiserver`) that should begin the bootstrapping of a new
+cluster for that application. This also applies at exit time. The Pilot is able
+to trap signals sent to the container and ensure the underlying application
+exits cleanly, and if it doesn't, report this diagnostic information back to
+the `navigator-apiserver`.
+
+Pilots should watch the `navigator-apiserver` for changes to resources it
+requires, and use the apiserver to communicate with other instances of Pilot
+within the same application cluster (eg. within the same Elasticsearch
+cluster). This may include things like notifications that a scale down event
+has been requested, or even used to elect a 'leader' between the Pilots in the
+event one is required for the proper functioning of that application.
+
+Each Pilot should also be able to expose a metrics endpoint to be scraped by
+systems like Prometheus. This could be implemented via some of the existing
+exporters for common applications, or directly within the Pilot itself if
+required.
+
+## navigator-controller-manager
+
+The `navigator-controller-manager` should implement all of the control features
+that are required for the proper functioning of the system. It should behave
+very much like the `kube-controller-manager`. If, for example, a user creates
+an `ElasticsearchCluster` resource, the `navigator-controller-manager` will
+create whatever corresponding resources are required in the target orchestrator
+in order to fulfill the request. It should then report status information of
+its actions back to the `navigator-apiserver`. Additional status information
+may also be provided by the Pilots themselves in this case, and in fact the
+`controller-manager` may also consequently perform *other* actions as a result.
+
+If a user were to modify the size of a node pool within an
+ElasticsearchCluster, the controller-manager would consequently notify the
+relevant Pilots of this event. It could then wait until a successful data
+migration has occured before actually scaling down the appropriate
+StatefulSet/Deployment. The precise semantics of *how* the controller-manager
+should communicate this to the Pilots is undefined, but however it is done it
+should be via the API server (either by updating the `status` field of the
+ElasticsearchCluster resource, or by creating an additional `ScaleEvent`
+resource to be fulfilled by the Pilots).
+
+Only one instance of the controller-manager should run at a given time, and
+they should leader-elect amongst themselves using either the
+`navigator-apiserver` or some other leader election primitive provided by the
+target orchestrator.
```

Comment:
(1) No, there is one controller manager for all at the moment, like how there is one controller manager for all k8s resource types. In future we could consider breaking this out into multiple standalone binaries, if there is a pressing reason to.

(2) Yes it is, and we *could* consider using it, however that may also put a dependence on running in an environment that already provides leader election primitives. By relying *only* on our own apiserver component, we can be sure that this will run outside of K8s/OS